nuxeo-drive-core/src/main/java/org/nuxeo/drive/service/impl/DefaultFileSystemItemFactory.java

```diff
@@ -137,16 +137,22 @@ public boolean isFileSystemItem(DocumentModel doc, boolean includeDeleted, boole
                     doc::getId, () -> e);
             return false;
         }
-        // Check Folderish or BlobHolder with a blob
-        if (!doc.isFolder() && blob == null) {
-            log.debug(
-                    "Document {} is not Folderish nor a BlobHolder with a blob, it cannot be adapted as a FileSystemItem.",
-                    doc::getId);
-            return false;
-        }
 
-        // Check for blobs backed by extended blob providers (ex: Google Drive)
         if (!doc.isFolder()) {
+            // Check Folderish or BlobHolder with a blob in the binary store
+            if (blob == null) {
+                log.debug(
+                        "Document {} is not Folderish nor a BlobHolder with a blob, it cannot be adapted as a FileSystemItem.",
+                        doc::getId);
+                return false;
+            } else if (blob.getDigest() == "notInBinaryStore") {
```

Comment:
That's ok for a hack for now but we'll have to clean this up better in the future. The "notInBinaryStore" digest of `MailMessageBlobHolder` is really bad.

Is there a way to add a unit test for that though? Because I think we could do a better test, based on `blob instanceof ManagedBlob` but we'd want to check that we do the correct thing for Notes, which use `DocumentStringBlobHolder` (called from `BlobHolderAdapterComponent.getBlobHolderAdapter`) which doesn't set any digest at all — but maybe that's ok for the Drive client.

Reply:
Yes this is definitely a hack. Drive would benefit from a more robust implementation since we often can't predict the behavior on new implementations.
IIRC the Drive client is okay with an empty digest.

Reply:
Should be `"notInBinaryStore".equals(blob.getDigest())` to test value equality (not reference) and avoid an NPE.

Agree to add a unit test, probably feasable with a Mock.

Also, the "notInBinaryStore" check could be part of the next block which already assumes that the blob is not null, and is kind of related (comment must be adapted then):
```
// Check for blobs backed by extended blob providers (ex: Google Drive)
if (!doc.isFolder()) {
    ....
}
```

Reply:
I already merged the two in my `if (!doc.isFolder()) {` block :)
I will be adding a test.

Reply:
Would it be better if it was
```
BlobHolder bh = doc.getAdapter(BlobHolder.class);
if (!doc.isFolder()) {
    if (blob == null) {
        // log.debug("Document has no blob ...
        return false;
    } else if (!(blob instanceof ManagedBlob or bh instanceof DocumentStringBlobHolder)) {
        // log.debug("Document can't be adapted as a FileSystemItem ...
        return false;
}
```
?

Reply:
Sounds good!

Reply:
Ok yes

Reply:
I'd prefer just this in the following block of tests (where we check the blob provider)

            } else if (blobProvider == null && !doc.getType().equals("Note")) {
                // not managed and not a Note
                return false;

Because Note is the only unmanaged blobs that are treated specially in the client